Add while loop config options and support custom-call root instruction in while loop unroll.

ptoulme-aws opened the pull request:
This PR adds the availability to configure while loop unroll thresholds. Existing defaults are maintained. This PR also adds the option for the user to specify an HloPassPipeline that will be run immediately before unroll. This is very important and by design. Some proprietary passes need to be run immediately before the unroll as the last step. The pipeline is defaulted to nullptr.

fhoushmand commented on xla/service/while_loop_unroller.cc:
```diff
@@ -759,6 +759,10 @@ absl::StatusOr<bool> WhileLoopUnroller::Run(
   VLOG(3) << "Number of while instructions in the module to unroll: "
           << unrollable_while_ops.size();
+  // Run pass pipeline specified by user immediately before unroll.
+  if (before_unroll_pipeline_) {
```
what's the purpose of this being inside the pass? any specific usecase?

ptoulme-aws replied:
We have certain use cases for our hardware that will insert custom-calls that need to be run inside this pass immediately before the unroll.

ptoulme-aws commented:
@ezhulenev can this be merged?

ptoulme-aws commented:
@xla-rotation  gentle ping

ptoulme-aws commented:
@fhoushmand can you help get this merged?

fhoushmand commented on xla/service/while_loop_unroller_test.cc:
```diff
@@ -1189,5 +1197,38 @@ TEST_F(WhileLoopUnrollerTest, IsEffectivelyStaticDynamicSlice) {
+
+  // Add a dead multiply instruction to the loop body.
+  Shape multiply_shape = induction_variable->shape();
+  HloInstruction* dead_multiply = loop_body->AddInstruction(
```
This variable is unused

fhoushmand replied:
@ptoulme-aws please fix this to merge the PR

ptoulme-aws replied:
Done

loislo commented on xla/service/while_loop_unroller.cc:
```diff
@@ -65,9 +65,9 @@ namespace {
 using hlo_query::ContainsInstrWithOpcode;
 
 // Parameters for the unroller that can be adjusted.
-const int kUnrollTripCountThreshold = 64;
```
I don't think you need this const values anymore.
Please use the properties of the UnrollConfig instead.

ptoulme-aws replied:
I did not do this as it requires routing the config through the pass which is significant changes to the pass.

loislo replied:
I don't think we want to have the global variables.

bchetioui replied:
This doesn't seem maintainable and creates hidden state for the `WhileLoopUnroller`. I don't think we should accept the PR in this state.

Can you explain what is the significant difficulty with passing it through the pass? Presumably, you already have access to this config through the `WhileLoopUnroller` itself, so it'd be a matter of passing it around only to helper functions---of which there seems to be `7` (assuming you need it everywhere). Or did I miss something?

loislo commented on xla/service/while_loop_unroller.cc:
```diff
@@ -737,7 +734,10 @@ absl::StatusOr<bool> WhileLoopUnroller::Run(
   bool changed = false;
-
+  // Allow user provided thresholds.
+  kUnrollTripCountThreshold = unroll_config_.trip_count_threshold;
```
ditto

ptoulme-aws replied:
Opted to leave as is to not route unroll_config through the pass which would be many changes.

loislo commented on xla/service/while_loop_unroller_test.cc:
```diff
@@ -1189,5 +1197,38 @@ TEST_F(WhileLoopUnrollerTest, IsEffectivelyStaticDynamicSlice) {
+  ASSERT_NE(induction_variable, nullptr);
+
+  // Add a dead multiply instruction to the loop body.
+  Shape multiply_shape = induction_variable->shape();
```
please add header for the Shape class.
I think it should be third_party/tensorflow/compiler/xla/shape.h

ptoulme-aws replied:
Done

ptoulme-aws commented:
@fhoushmand I made the changes you requested. Can we merge now? Thank you

bchetioui commented:
Copying it here because GitHub seems intent on leaving the comment resolved:

Th[e unconstification of global constants) doesn't seem maintainable and creates hidden state for the `WhileLoopUnroller`. I don't think we should accept the PR in this state.

Can you explain what is the significant difficulty with passing it through the pass? Presumably, you already have access to this config through the `WhileLoopUnroller` itself, so it'd be a matter of passing it around only to helper functions---of which there seems to be `7` (assuming you need it everywhere). Or did I miss something?

ptoulme-aws commented:
> Copying it here because GitHub seems intent on leaving the comment resolved:
> 
> Th[e unconstification of global constants) doesn't seem maintainable and creates hidden state for the `WhileLoopUnroller`. I don't think we should accept the PR in this state.
> 
> Can you explain what is the significant difficulty with passing it through the pass? Presumably, you already have access to this config through the `WhileLoopUnroller` itself, so it'd be a matter of passing it around only to helper functions---of which there seems to be `7` (assuming you need it everywhere). Or did I miss something?

I have made the changes you requested. I initially did not do it because some of the methods were static. I have made all the changes now.

bchetioui commented on xla/service/while_loop_unroller.cc:
```diff
@@ -226,42 +221,39 @@ UnrollSingleIterationOfTrivialLoop(HloInstruction* while_op,
 // 3. unroll expansion limit (#_body_instructions * trip_count).
 // These conditions can be changed per usecase.
-bool InitialFeasibilityCheck(HloInstruction* while_op, WhileLoopConfig config) {
+bool InitialFeasibilityCheck(HloInstruction* while_op, WhileLoopConfig config, UnrollConfig unroll_config) {
```
nit: seems like this could be a const ref?

fhoushmand replied:
@ptoulme-aws please fix this

bchetioui commented on xla/service/while_loop_unroller_test.cc:
```diff
@@ -1189,5 +1198,37 @@ TEST_F(WhileLoopUnrollerTest, IsEffectivelyStaticDynamicSlice) {
   }
 }
 
+TEST_F(WhileLoopUnrollerTest, SimpleLoopWithPassPipeline) {
```
The name doesn't make it obvious what is being run here. Perhaps rename to `WhileLoopUnrollerExecutesCallbackPassPipeline`?

Note that it also doesn't check *when* that pass pipeline is run. Maybe that's OK (?).

ptoulme-aws replied:
Pass pipeline deleted.

bchetioui commented on xla/service/while_loop_unroller.cc:
```diff
@@ -755,10 +746,14 @@ absl::StatusOr<bool> WhileLoopUnroller::Run(
   VLOG(3) << "Number of while instructions in the module to unroll: "
           << unrollable_while_ops.size();
+  // Run pass pipeline specified by user immediately before unroll.
+  if (before_unroll_pipeline_) {
```
Logically, this seems like a preparation step for the module before unrolling.

I think it might make more sense to execute this pipeline as part of `PrepareModuleForUnrolling` (e.g. at the end of it), as opposed to in the middle of the pass. I think it'd have the same effect as what you're trying to achieve here, but might separate the logic more cleanly (at the cost of adding another parameter to a static method, which seems OK). WDYT?

This also improves testability a little bit, in the sense that you could now call the `PrepareModuleForUnrolling` static method directly in your test to verify that preparation calls your pipeline.

ptoulme-aws replied:
Certain passes that are proprietary to our hardware cannot be run there. They have to be run at the location I specified. Please see my discussion earlier with another reviewer about this.

bchetioui replied:
> Certain passes that are proprietary to our hardware cannot be run there. They have to be run at the location I specified. Please see my discussion earlier with another reviewer about this.

I saw this discussion, but I didn't buy that argument at face value (and I obviously don't know your passes): nothing happens between the call to `PrepareModuleForUnrolling`  and this conditional, beyond

1. finding all the while ops in the module, and
2. finding all the while ops we think we can unroll.

For your passes to cause a difference there would mean that your proprietary passes (a) insert new while ops that you *don't* want to process, or that (b) they remove/modify while ops that you wanted to (somehow) aggregate in these vectors. (b) doesn't seem like something one'd want to do at all, and (a) seems like something that could be done by inserting the while ops after the `WhileLoopUnroller` pass has run, or before by inserting loops that can not be unrolled.

For any other use case, it seems like it should make no practical difference whether this pass pipeline is called here or as the last step in `PrepareModuleForUnrolling`. However, in terms of code organization, it makes much more sense to me for this pipeline to be called there.

Am I missing something? Are you truly sure that putting it here should make a difference compared to putting this at the end of `PrepareModuleForUnrolling`?

ptoulme-aws replied:
Yes, I am sure. For some of our proprietary passes if you run them before WhileLoopUnroller then while loop unroller will say the body cannot be unrolled. This is due to some of the transformations we make. The body can still be unrolled so thats why we want to run the passes directly before we call the unroll function.

bchetioui replied:
> Yes, I am sure. For some of our proprietary passes if you run them before WhileLoopUnroller then while loop unroller will say the body cannot be unrolled. This is due to some of the transformations we make. The body can still be unrolled so thats why we want to run the passes directly before we call the unroll function.

Just to make sure we didn't misunderstand each other, note that I wasn't suggesting actually running your pass pipeline before the`WhileLoopUnroller`, but at the end of `PrepareModuleForUnrolling` (which is firmly inside the `WhileLoopUnroller` pass, and called just a few lines above).

ptoulme-aws replied:
Sorry I meant PrepareModuleForUnrolling. If you run it in PrepareModuleForUnrolling, GetUnrollableLoops will say the loops cannot be unrolled. This is because some of our transformations break those conditions. Hence we run the transformation immediately before the unroll call itself. 
https://github.com/openxla/xla/blob/d5d9a751e03fc147c5c64160feb7602f6348eb6c/xla/service/while_loop_unroller.cc#L749

fhoushmand replied:
I'm gonna suggest that you look at why the PrepareModuleForUnrolling is not sufficient to make the loops in question (the ones that got transformed by your passes) unrollable.

ptoulme-aws replied:
Without explaining too much, these proprietary passes insert custom-calls inside the body of the while loop. Why does GetUnrollableLoops fail if there are custom-calls in the body of the body region? I can try and debug that, but if that is by design I do not understand why providing power users (hardware vendors) the option to control exactly where they want to run some passes for unrolling. This change is mainly for power users of XLA, so I do not understand the pushback.

fhoushmand replied:
The reason is that this particular design choice might be an artifact of a more serious invariant that is breaking here and I overlooked earlier. I think the fact that GetUnrollableLoops fails if there are custom-calls added to the while body, by itself is a red flag. These custom-calls might be in fact the reason that make the loop unrollable (e.g, they have side effect and are inside the condition which we end up removing after unrolling). I suggest that you look at the IsLoopUnrollable function and see if those custom-calls are violating any conditions there. I believe after this investigation we are able to better decide how to move forward.

ptoulme-aws replied:
OK I can investigate this. The loop is definitely unrollable with the custom-calls, as the custom-calls are no ops. They are our proprietary custom-calls but have no real function.

ptoulme-aws replied:
I have reduced a very small test case that can trigger the unroller to refuse to unroll.

ptoulme-aws replied:
Found the problem. custom-calls at start and end of the loop break 
  std::optional<int64_t> indvar_tuple_idx =
      GetLoopInductionVarTupleIdx(while_op);
  if (!indvar_tuple_idx.has_value()) {
    LOG(INFO) << "Break on line 582";
    return std::nullopt;
  }

ptoulme-aws replied:
This is the problem @fhoushmand @bchetioui 
```
I0000 00:00:1723832752.201489   52013 while_loop_analysis.cc:293] While body's root is not a tuple instruction: %custom-call.2 = (s32[]{:T(128)}, s32[3]{0}) custom-call((s32[]{:T(128)}, s32[3]{0}) %tuple), custom_call_target="CustomCallEnd"
```

Can I add an optional config to GetLoopInductionVarTupleIdx saying allow_custom_calls=True? Now you see why I wanted to run the pass pipeline immediately before unroll time as some passes break the induction var pattern match. I still think my original approach is preferred.

ptoulme-aws replied:
Reproducible test case 
```
TEST_F(WhileLoopUnrollerTest, SimpleLoopWithCustomCall){
  std::string hlo_string = R"(
  HloModule SimpleLoop
  SimpleLoop.body {
    loop_var.1 = (s32[]{:T(128)}, s32[3]{0}) parameter(0)
    custom-call.1 = (s32[]{:T(128)}, s32[3]{0}) custom-call(loop_var.1), custom_call_target="CustomCallStart"
    get-tuple-element.1 = s32[]{:T(128)} get-tuple-element(custom-call.1), index=0
    constant.1 = s32[]{:T(128)} constant(1)
    idx = s32[]{:T(128)} add(get-tuple-element.1, constant.1)
    get-tuple-element.2 = s32[3]{0} get-tuple-element(custom-call.1), index=1
    output = s32[3]{0} add(get-tuple-element.2, get-tuple-element.2)
    tuple = (s32[]{:T(128)}, s32[3]{0}) tuple(idx, output)
    ROOT custom-call.2 = (s32[]{:T(128)}, s32[3]{0}) custom-call(tuple), custom_call_target="CustomCallEnd"
  }
  SimpleLoop.condition {
    loop_var.2 = (s32[]{:T(128)}, s32[3]{0}) parameter(0)
    get-tuple-element.3 = s32[] get-tuple-element(loop_var.2), index=0
    constant.2 = s32[]{:T(128)} constant(5)
    ROOT less-than = pred[] compare(get-tuple-element.3, constant.2), direction=LT
  }
  ENTRY SimpleLoop {
    constant.3 = s32[]{:T(128)} constant(0)
    constant.4 = s32[3]{0} constant({0, 1, 2})
    tuple.1 = (s32[]{:T(128)}, s32[3]{0}) tuple(constant.3, constant.4)
    ROOT while = (s32[]{:T(128)}, s32[3]{0}) while(tuple.1), condition=
      SimpleLoop.condition, body=SimpleLoop.body
  }
  )";
  auto m = ParseAndReturnVerifiedModule(hlo_string).value();
  UnrollConfig config;
  EXPECT_TRUE(WhileLoopUnroller(/*unroll_factor=*/-1, /*wrap_in_trivial_loop=*/false, config).Run(m.get()).value());


}
```

fhoushmand replied:
Thanks @ptoulme-aws for looking into this.

> Can I add an optional config to GetLoopInductionVarTupleIdx saying allow_custom_calls=True? 
Caller should not care about what kind of instructions are in the while body. If we can get the induction var statically at compile time, that function should be able to handle it without needing to pass additional arguments. The reason that this was not handled in the function is this pattern was never observed. You can add it though.

> Now you see why I wanted to run the pass pipeline immediately before unroll time as some passes break the induction var pattern match. I still think my original approach is preferred.

I'm gonna have to disagree with you on that. Perhaps, your approach might be easier, but it would be dangerous. Assume that the pipeline is given a pass that can change the trip count of the loops. It's easy to see that doing surgery in the middle of the pass would break invariants that the pass depends on.

I like to see each compiler pass as an atomic unit that operates on a module that satisfies a pre-condition, and produces a module satisfying a post-condition. Given this, one must be cautious when breaking atomicity.

ptoulme-aws replied:
Thank you. I will add the pattern and remove the interior pass pipeline I added and add tests for new patterns.

bchetioui reviewed (changes requested):
Thanks for the update! A couple more small comments.

ptoulme-aws commented:
@fhoushmand and @bchetioui I have removed the intermediate pass pipeline, and I have added support for custom-call as a root instruction for unrolling. I have added tests in both the while loop unroller and while loop analysis targets. 
Thank you!

fhoushmand commented on xla/service/while_loop_analysis.cc:
```diff
@@ -289,13 +290,37 @@ optional<int64_t> GetLoopInductionVarTupleIdx(const HloInstruction* while_op) {
   // where while_body_inc is operand N of while_body_root.
```
Please update the comment according to your changes that support custom-call

ptoulme-aws commented:
@fhoushmand comments resolved.

ptoulme-aws commented:
@fhoushmand @bchetioui can we merge now?

sgerrard commented:
I am getting an internal error:

[third_party/tensorflow/compiler/xla/service/while_loop_unroller_test.cc:35](https://cs.corp.google.com/piper///depot/google3/third_party/tensorflow/compiler/xla/service/while_loop_unroller_test.cc?l=35&ws=tap-presubmit-rerun-server/39310130&snapshot=2):10: error: module //third_party/tensorflow/compiler/xla/service:while_loop_unroller_test does not depend on a module exporting 'third_party/tensorflow/compiler/xla/shape.h'
see [http://go/cpp-features#layering_check](https://www.google.com/url?q=http://go/cpp-features%23layering_check&sa=D); to fix run:
	build_cleaner //third_party/tensorflow/compiler/xla/service:while_loop_unroller_test
   35 | #include "third_party/tensorflow/compiler/xla/shape.h"
      |          ^
1 error generated.

Can you take a look? Perhaps update the build file for the tests?

ptoulme-aws commented:
@sgerrard just pushed the change to remove that line. It was dead import. Thanks!

ptoulme-aws commented:
@sgerrard is the PR ok now?

fhoushmand commented:
> @sgerrard is the PR ok now?

yes, it should get merged soon

sgerrard commented:
It's merging now.

sgerrard commented:
The submit is block due to out of date files, can you resolve conflicts? @ptoulme-aws

ptoulme-aws commented:
@sgerrard I rebased to TOT and fixed the conflicts.

shraiysh commented on xla/service/while_loop_analysis_test.cc:
```diff
@@ -130,6 +130,71 @@ TEST_F(WhileLoopAnalysisTest, SingleIterationUpperBound) {
+  HloModule SimpleLoop
+  SimpleLoop.body {
+    loop_var.1 = (s32[]{:T(128)}, s32[3]{0}) parameter(0)
+    custom-call.1 = (s32[]{:T(128)}, s32[3]{0}) custom-call(loop_var.1), custom_call_target="CustomCallStart"
```
Here, are we assuming that the custom calls at the beginning and end do not modify the loop trip count (especially decrement it)? This seems a bit odd. What is the rationale behind this?

ptoulme-aws replied:
Correct they will not modify the tripcount. 
We are guaranteed that also, as these custom-calls are not modifying the while loop trip count logic. If for some reason, the trip count was modified - it would refuse to unroll as static trip count fails. 

You can read the rationale for this in the PR. Some while loop bodies we still want to unroll that have custom-calls at start and end of the while loop body.

ptoulme-aws replied:
So in while loop analysis it calls ComputeWhileLoopTripCount (function may be named differently but similar) this computes the trip count of the while loop and if it is static then it allows an unroll.

akuegel replied:
I think the assumption here is that the custom call does not use the passed in loop induction variable in-place. We cannot know what a custom call does, but as long as it does not have in-place parameters, it cannot modify the passed in parameter. To make a custom call have in-place parameters, it needs to be explicitly handled in buffer assignment (e.g. here: https://github.com/openxla/xla/blob/0fc891390264fb85ac822f45c4106c48e1a10ffc/xla/service/gpu/buffer_sharing.cc#L232).

shraiysh replied:
The problem here is that we are using the output of this custom call to update it and then we send that value as the updated induction variable. It doesn't matter if the update is in-place or not, because we are using the output of the custom call for the increment operation. For this to be safe, we need to somehow assert that the custom call returns the loop induction variable unchanged. Am I missing something here?

akuegel replied:
Ah, good point, I had not noticed that. I agree, this seems problematic. In fact, if the custom call returns the induction variable unchanged, why make it an output of the custom call in the first place?

ptoulme-aws replied:
So I remember now - it is assumed the custom-call will not alter the trip count. This is a special case for Neuron hardware in that we sometimes have custom-calls at the start and end of our while loop body, and we want to still unroll them.

shraiysh replied:
A safer solution to this problem in my opinion would to be to add an unsafe pass `replace-while-custom-call-ind-var` right before unrolling (default to off). This pass would replace the uses of the returned induction variable by the original induction variable. This pass would only run for the special case of neuron hardware and will not make the while loop analysis unsafe. For this testcase, the pass would change the HLO to the following:
```
SimpleLoop.body {
    loop_var.1 = (s32[]{:T(128)}, s32[3]{0}) parameter(0)
    custom-call.1 = (s32[]{:T(128)}, s32[3]{0}) custom-call(loop_var.1), custom_call_target="CustomCallStart"
    get-tuple-element.1 = s32[]{:T(128)} get-tuple-element(loop_var.1), index=0
    constant.1 = s32[]{:T(128)} constant(1)
    idx = s32[]{:T(128)} add(get-tuple-element.1, constant.1)
    get-tuple-element.2 = s32[3]{0} get-tuple-element(custom-call.1), index=1
    output = s32[3]{0} add(get-tuple-element.2, get-tuple-element.2)
    custom-call.2 = (s32[]{:T(128)}, s32[3]{0}) custom-call(idx, output), custom_call_target="CustomCallEnd"
    final_output_data = s32[3]{0} get-tuple-element(custom-call.2), index=1
    ROOT tuple = (s32[]{:T(128)}, s32[3]{0})tuple(idx, final_output_data)
  }
```
With such a pass, while loop unrolling should still unroll the loop. Will this work for you @ptoulme-aws ? What do you think about this solution @akuegel @fhoushmand ?

ptoulme-aws replied:
For my use case, I needed the custom-calls to be the first and last instructions of the while loop body, so this would not work.

I wanted to be able to just run the pass that inserts these custom-calls (proprietary pass) immediately before unroll but that PR was not accepted. Perhaps you could add a boolean flag - allow custom-call induction var?

shraiysh replied:
We should not make the while analysis unsafe (even behind a flag) because many passes rely on this pass and the trip count calculated. If the proprietary passes require the custom calls to be the first and last instructions to be custom-calls, then maybe there should be passes for both these tasks:

1. Before unroll pass, one pass should replace the uses of output of custom call to the original loop induction variable (as suggested above)
2. After unroll pass, another pass should start reusing the output of custom call as the loop induction variable and make the custom-calls the first and last instructions of each iteration of the unrolled loop so that the proprietary passes can run on them.

Both these passes would be internal (or default to off in upstream), and so this would not make while loop analysis or unrolling inaccurate upstream.

ptoulme-aws replied:
How is a flag unsafe if the caller guarantees the custom-calls are just pass through for the trip count var? 
Those solutions are all cumbersome. A flag is a simple solution.

ptoulme-aws replied:
You can default flag to off obviously. My point is - it is not unsafe if the person using the flags know why they are using it. 
Default to False.

bchetioui replied:
Thanks for catching this issue, I'm ashamed that I didn't notice it!

> How is a flag unsafe if the caller guarantees the custom-calls are just pass through for the trip count var? Those solutions are all cumbersome. A flag is a simple solution.
>
> How is a flag unsafe if the caller guarantees the custom-calls are just pass through for the trip count var?
Those solutions are all cumbersome. A flag is a simple solution.

We do not generally wish to support flags in XLA, unless there's a particularly good reason to do that—and this does not seem like one. Flags are cumbersome, and significantly increase our testing & maintenance surface area.

The solution you suggest @shraiysh sounds reasonable. If there is no way to make this safe in the general case, and we insist that Neuron needs to specialize this use case, then the dangerous transformation needs to be gated for Neuron hardware only.

Failing that, we'll probably need to revert this, since the code as it is right now is unsafe.

shraiysh replied:
Do we have the steps forward? Do we have a design for a safe implementation of this in while loop analysis?

ptoulme-aws replied:
Looking at your proposal 

1. Before unroll pass, one pass should replace the uses of output of custom call to the original loop induction variable (as suggested above)
2. After unroll pass, another pass should start reusing the output of custom call as the loop induction variable and make the custom-calls the first and last instructions of each iteration of the unrolled loop so that the proprietary passes can run on them.

ptoulme-aws replied:
For 1. This assumes custom-call does not change the loop induction var. This is effectively same assumption as the current pass.

ptoulme-aws replied:
2. This also assumes it does not change loop induction var.

ptoulme-aws replied:
Your proposal is effectively "if loop induction var is output of custom-call, change graph to use loop induction var of custom-call for unroll purposes to force a guranteed static trip count, but then put the loop induction var back in the custom-call."

Am I understanding correctly? This does not seem safe and is not transparent to users

shraiysh replied:
The core transform that neuron hardware requires here (according to the description above) is unsafe. In the proposed solution, the unsafe part of the transform is decoupled from the while loop analysis and does not affect the compilation for other hardware and users. These two unsafe passes would be internal to neuron hardware. 

We do not have to solve it with this approach I outlined above, but irrespective of the approach we resort to, we need to make while loop analysis safe, which it isn't in its current state. 

Meanwhile, I suggest reverting this PR to make while loop analysis safe.

ptoulme-aws replied:
At this point, perhaps PRing the custom-call transform is not realistic. Neuron will have to just continue to use internal while loop unroller. 

@shraiysh you can revert the PR. Perhaps you can keep the while loop config PR portion as that is quite useful.
